src/BinomialQueue/Max.hs

```diff
@@ -0,0 +1,292 @@
+dropWhile p = MaxQueue . MinQ.dropWhile (p . unDown) . unMaxQueue
+
+-- | 'span', applied to a predicate @p@ and a queue @queue@, returns a tuple where
+-- first element is longest prefix (possibly empty) of @queue@ of elements that
+-- satisfy @p@ and second element is the remainder of the queue.
+span :: Ord a => (a -> Bool) -> MaxQueue a -> ([a], MaxQueue a)
+span p (MaxQueue queue)
+  | (front, rear) <- MinQ.span (p . unDown) queue
+  = (fmap unDown front, MaxQueue rear)
+
+-- | 'break', applied to a predicate @p@ and a queue @queue@, returns a tuple where
+-- first element is longest prefix (possibly empty) of @queue@ of elements that
+-- /do not satisfy/ @p@ and second element is the remainder of the queue.
+break :: Ord a => (a -> Bool) -> MaxQueue a -> ([a], MaxQueue a)
+break p = span (not . p)
+
+{-# INLINE take #-}
+-- | /O(k log n)/. 'take' @k@, applied to a queue @queue@, returns a list of the greatest @k@ elements of @queue@,
+-- or all elements of @queue@ itself if @k >= 'size' queue@.
+take :: Ord a => Int -> MaxQueue a -> [a]
+take n = List.take n . toDescList
+
+-- | /O(k log n)/. 'drop' @k@, applied to a queue @queue@, returns @queue@ with the greatest @k@ elements deleted,
+-- or an empty queue if @k >= size 'queue'@.
+drop :: Ord a => Int -> MaxQueue a -> MaxQueue a
+drop n (MaxQueue queue) = MaxQueue (MinQ.drop n queue)
+
+-- | /O(k log n)/. Equivalent to @('take' k queue, 'drop' k queue)@.
+splitAt :: Ord a => Int -> MaxQueue a -> ([a], MaxQueue a)
+splitAt n (MaxQueue queue)
+  | (l, r) <- MinQ.splitAt n queue
+  = (fmap unDown l, MaxQueue r)
+
+-- | /O(n)/. Returns the queue with all elements not satisfying @p@ removed.
+filter :: Ord a => (a -> Bool) -> MaxQueue a -> MaxQueue a
+filter p = MaxQueue . MinQ.filter (p . unDown) . unMaxQueue
+
+-- | /O(n)/. Returns a pair where the first queue contains all elements satisfying @p@, and the second queue
+-- contains all elements not satisfying @p@.
+partition :: Ord a => (a -> Bool) -> MaxQueue a -> (MaxQueue a, MaxQueue a)
+partition p = go . unMaxQueue
+  where
+    go queue
+      | (l, r) <- MinQ.partition (p . unDown) queue
+      = (MaxQueue l, MaxQueue r)
+
+-- | /O(n)/. Creates a new priority queue containing the images of the elements of this queue.
+-- Equivalent to @'fromList' . 'Data.List.map' f . toList@.
+map :: Ord b => (a -> b) -> MaxQueue a -> MaxQueue b
+map f = MaxQueue . MinQ.map (fmap f) . unMaxQueue
+
+{-# INLINE toList #-}
+-- | /O(n log n)/. Returns the elements of the priority queue in descending order. Equivalent to 'toDescList'.
+--
+-- If the order of the elements is irrelevant, consider using 'toListU'.
+toList :: Ord a => MaxQueue a -> [a]
+toList = fmap unDown . MinQ.toAscList . unMaxQueue
+
+toAscList :: Ord a => MaxQueue a -> [a]
+toAscList = fmap unDown . MinQ.toDescList . unMaxQueue
+
+toDescList :: Ord a => MaxQueue a -> [a]
+toDescList = fmap unDown . MinQ.toAscList . unMaxQueue
```

Comment:
These need doc comments. Same for a bunch of other functions below.

Reply:
Ugh, yes. Looks like I need lots of boring documentation work.